consensus/src/transaction_shuffler/fairness/conflict_key.rs

```diff
@@ -0,0 +1,175 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::transaction_shuffler::fairness::NUM_CONFLICT_ZONES;
+use aptos_types::transaction::{SignedTransaction, TransactionPayload};
+use move_core_types::{
+    account_address::AccountAddress, identifier::Identifier, language_storage::ModuleId,
+};
+use std::{collections::HashMap, hash::Hash};
+
+pub(crate) trait ConflictKey: Eq + Hash + PartialEq {
+    fn extract_from(txn: &SignedTransaction) -> Self;
+
+    fn conflict_exempt(&self) -> bool {
+        false
+    }
+}
+
+#[derive(Eq, Hash, PartialEq)]
+struct TxnSenderKey(AccountAddress);
+
+impl ConflictKey for TxnSenderKey {
+    fn extract_from(txn: &SignedTransaction) -> Self {
+        TxnSenderKey(txn.sender())
+    }
+}
+
+#[derive(Eq, Hash, PartialEq)]
+enum EntryFunModuleKey {
+    Module(ModuleId),
+    AnyScriptOrModuleBundle,
+    AnyMultiSig,
+    Exempt,
+}
+
+impl ConflictKey for EntryFunModuleKey {
+    fn extract_from(txn: &SignedTransaction) -> Self {
+        match txn.payload() {
+            TransactionPayload::EntryFunction(entry_fun) => {
+                Self::Module(entry_fun.module().clone())
+            },
+            TransactionPayload::Multisig(..) => Self::AnyMultiSig,
```

Comment:
We can also get the EntryFunction from within the MultiSig so it doesn't fall in a single bucket (as suggested by @gregnazario in my PFN PR): https://github.com/aptos-labs/aptos-core/blob/075855844b2b53ce4152f095becd9e12db2d7556/api/src/transactions.rs#L925

Reply:
Yeah will need to follow up. (trying to under what happens when the payload is None. )